### ISSUE TITLE
Discrepancy between `Array#[]` and `Slice#[]`

### ISSUE THREAD

**Issue by asterite** [crystal-lang/crystal#12617]
## Bug Report

This is for Array:

```crystal
a = [1]
p a[0..10] # => [1]
```

This is for Slice:

```crystal
bytes = Bytes[1]
p bytes[0..10] # raises IndexError
```

I don't think the second case should raise. When you take a subslice it should be of at most the range you say, and it should never fail unless the first index is out of bounds (like Array, I think)

**Comment by HertzDevil** [comment]
finally someone says it O_O

**Comment by caspiano** [comment]
Methods that currently accept `Range` arguments that currently raise when the end is out of bounds

- `String#sub(range : Range, replacement : Char)`
- `String#sub(range : Range, replacement : String)`
- `Slice#[](range : Range)`
- `Slice#[]?(range : Range)`

@HertzDevil There is no `Indexable#[](range : Range)`, so I don't think this is necessarily an inconsistency in `Indexable` unless its explicitly added to the module.

**Comment by straight-shoota** [comment]
Even though it's not part of the actual `Indexable` module, it's kind of an extended `Indexable` interface that many implementations follow (but not necessarily all, hence it's not part of the module).
There is certainly a high similarity between different implementations of these methods in types that include `Indexable` and it makes sense for them to have the same error behaviour.

**Comment by beta-ziliani** [comment]
But there could be a default implementation in `Indexable`, right?

**Comment by straight-shoota** [comment]
No. There's no generic representation for a subsequence of a specific collection type. And this might not even be a legal operation, because `Indexable` is a purely a read interface and there's no guarantee that you could actually create an instance of an implementing type to represent a subsequence.
